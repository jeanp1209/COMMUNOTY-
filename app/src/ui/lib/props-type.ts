// based on https://codewithstyle.info/TypeScript-conditional-types-real-life-example/
export type PropsType<C> = C extends React.Component<infer P>
  ? P
  : C extends React.SFC<infer P>
  ? P
  : never
