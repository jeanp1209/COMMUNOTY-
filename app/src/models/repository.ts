import * as Path from 'path'

import { GitHubRepository } from './github-repository'
import { IAheadBehind } from './branch'

function getBaseName(path: string): string {
  const baseName = Path.basename(path)

  if (baseName.length === 0) {
    // the repository is at the root of the drive
    // -> show the full path here to show _something_
    return path
  }

  return baseName
}

/** A local repository. */
export class Repository {
  public readonly name: string

  /**
   * @param path The working directory of this repository
   * @param missing Was the repository missing on disk last we checked?
   */
  public constructor(
    public readonly path: string,
    public readonly id: number,
    public readonly gitHubRepository: GitHubRepository | null,
    public readonly missing: boolean
  ) {
    this.name = (gitHubRepository && gitHubRepository.name) || getBaseName(path)
  }

  /**
   * A hash of the properties of the object.
   *
   * Objects with the same hash are guaranteed to be structurally equal.
   */
  public get hash(): string {
    return `${this.id}+${this.gitHubRepository && this.gitHubRepository.hash}+${
      this.path
    }+${this.missing}+${this.name}`
  }
}

/**
 * A snapshot for the local state for a given repository
 */
export interface ILocalRepositoryState {
  /**
   * The ahead/behind count for the current branch, or `null` if no tracking
   * branch found.
   */
  readonly aheadBehind: IAheadBehind | null
  /**
   * The number of uncommitted changes currently in the repository.
   */
  readonly changedFilesCount: number
}

/**
 * Returns the owner/name alias if associated with a GitHub repository,
 * otherwise the folder name that contains the repository
 */
export function nameOf(repository: Repository) {
  const { gitHubRepository } = repository

  return gitHubRepository !== null ? gitHubRepository.fullName : repository.name
}
