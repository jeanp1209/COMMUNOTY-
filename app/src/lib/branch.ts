import { Branch, BranchType } from '../models/branch'
import { UpstreamRemoteName } from './stores'

/**
 * Finds the remote branch for a branch in the upstream repository
 *
 * For example:
 * If official/funnel has the branch `development`, then running
 * this on the branches from the fork outofambit/funnel with
 * `development` as the specified branch name will find the
 * branch `remotes/upstream/development`
 *
 * @param branchName short name of the branch in the upstream repo
 * @param branches all the branches in the local repo
 */
export function findUpstreamRemoteBranch(
  branchName: string,
  branches: ReadonlyArray<Branch>
) {
  return branches.find(
    b =>
      b.type === BranchType.Remote &&
      b.name === `${UpstreamRemoteName}/${branchName}`
  )
}
