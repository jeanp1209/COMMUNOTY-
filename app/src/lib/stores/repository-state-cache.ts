import { Branch } from '../../models/branch'
import { Commit } from '../../models/commit'
import { PullRequest } from '../../models/pull-request'
import { Repository } from '../../models/repository'
import {
  CommittedFileChange,
  WorkingDirectoryFileChange,
  WorkingDirectoryStatus,
} from '../../models/status'
import { TipState } from '../../models/tip'
import {
  ComparisonView,
  IBranchesState,
  IChangesState,
  ICompareState,
  IRepositoryState,
  RepositorySectionTab,
} from '../app-state'
import { ComparisonCache } from '../comparison-cache'
import { IGitHubUser } from '../databases'
import { merge } from '../merge'

export class RepositoryStateCache {
  private readonly repositoryState = new Map<string, IRepositoryState>()

  public constructor(
    private readonly getUsersForRepository: (
      repository: Repository
    ) => Map<string, IGitHubUser>
  ) {}

  /** Get the state for the repository. */
  public get(repository: Repository): IRepositoryState {
    const existing = this.repositoryState.get(repository.hash)
    if (existing != null) {
      const gitHubUsers = this.getUsersForRepository(repository)
      return merge(existing, { gitHubUsers })
    }

    const newItem = getInitialRepositoryState()
    this.repositoryState.set(repository.hash, newItem)
    return newItem
  }

  public update<K extends keyof IRepositoryState>(
    repository: Repository,
    fn: (state: IRepositoryState) => Pick<IRepositoryState, K>
  ) {
    const currentState = this.get(repository)
    const newValues = fn(currentState)
    this.repositoryState.set(repository.hash, merge(currentState, newValues))
  }

  public updateCompareState<K extends keyof ICompareState>(
    repository: Repository,
    fn: (state: ICompareState) => Pick<ICompareState, K>
  ) {
    this.update(repository, state => {
      const compareState = state.compareState
      const newValues = fn(compareState)

      return { compareState: merge(compareState, newValues) }
    })
  }

  public updateChangesState<K extends keyof IChangesState>(
    repository: Repository,
    fn: (changesState: IChangesState) => Pick<IChangesState, K>
  ) {
    this.update(repository, state => {
      const changesState = state.changesState
      const newState = merge(changesState, fn(changesState))
      return { changesState: newState }
    })
  }

  public updateBranchesState<K extends keyof IBranchesState>(
    repository: Repository,
    fn: (branchesState: IBranchesState) => Pick<IBranchesState, K>
  ) {
    this.update(repository, state => {
      const changesState = state.branchesState
      const newState = merge(changesState, fn(changesState))
      return { branchesState: newState }
    })
  }
}

function getInitialRepositoryState(): IRepositoryState {
  return {
    commitSelection: {
      sha: null,
      file: null,
      changedFiles: new Array<CommittedFileChange>(),
      diff: null,
    },
    changesState: {
      workingDirectory: WorkingDirectoryStatus.fromFiles(
        new Array<WorkingDirectoryFileChange>()
      ),
      selectedFileIDs: [],
      diff: null,
      contextualCommitMessage: null,
      commitMessage: null,
      coAuthors: [],
      showCoAuthoredBy: false,
    },
    selectedSection: RepositorySectionTab.Changes,
    branchesState: {
      tip: { kind: TipState.Unknown },
      defaultBranch: null,
      allBranches: new Array<Branch>(),
      recentBranches: new Array<Branch>(),
      openPullRequests: new Array<PullRequest>(),
      currentPullRequest: null,
      isLoadingPullRequests: false,
    },
    compareState: {
      isDivergingBranchBannerVisible: false,
      formState: { kind: ComparisonView.None },
      mergeStatus: null,
      showBranchList: false,
      filterText: '',
      commitSHAs: [],
      aheadBehindCache: new ComparisonCache(),
      allBranches: new Array<Branch>(),
      recentBranches: new Array<Branch>(),
      defaultBranch: null,
      inferredComparisonBranch: { branch: null, aheadBehind: null },
    },
    commitAuthor: null,
    gitHubUsers: new Map<string, IGitHubUser>(),
    commitLookup: new Map<string, Commit>(),
    localCommitSHAs: [],
    aheadBehind: null,
    remote: null,
    isPushPullFetchInProgress: false,
    isCommitting: false,
    lastFetched: null,
    checkoutProgress: null,
    pushPullFetchProgress: null,
    revertProgress: null,
  }
}
