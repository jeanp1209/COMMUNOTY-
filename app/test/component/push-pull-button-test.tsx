import * as React from 'react'
import * as moment from 'moment'

import { render, cleanup } from 'react-testing-library'
import { PushPullButton } from '../../src/ui/toolbar/push-pull-button'
import { InMemoryDispatcher } from '../helpers/in-memory-dispatcher'
import {
  AppStore,
  GitHubUserStore,
  IssuesStore,
  CloningRepositoriesStore,
  SignInStore,
  RepositoriesStore,
  AccountsStore,
  PullRequestStore,
} from '../../src/lib/stores'
import { ApiRepositoriesStore } from '../../src/lib/stores/api-repositories-store'
import { RepositoryStateCache } from '../../src/lib/stores/repository-state-cache'
import {
  TestGitHubUserDatabase,
  TestIssuesDatabase,
  TestStatsDatabase,
  TestRepositoriesDatabase,
  TestPullRequestDatabase,
} from '../helpers/databases'
import { TestActivityMonitor } from '../helpers/test-activity-monitor'
import { StatsStore } from '../../src/lib/stats'
import { AsyncInMemoryStore, InMemoryStore } from '../helpers/stores'
import { TipState } from '../../src/models/tip'
import { setupEmptyRepository } from '../helpers/repositories'
import { IFetchProgress } from '../../src/models/progress'
import { Repository } from '../../src/models/repository'
import { PropsType } from '../../src/ui/lib/props-type'

describe('<PushPullButton/>', () => {
  let defaultProps: PropsType<PushPullButton>
  let repository: Repository

  afterEach(cleanup)

  beforeEach(async () => {
    const db = new TestGitHubUserDatabase()
    await db.reset()

    const issuesDb = new TestIssuesDatabase()
    await issuesDb.reset()

    const statsDb = new TestStatsDatabase()
    await statsDb.reset()
    const statsStore = new StatsStore(statsDb, new TestActivityMonitor())

    const repositoriesDb = new TestRepositoriesDatabase()
    await repositoriesDb.reset()
    const repositoriesStore = new RepositoriesStore(repositoriesDb)

    const accountsStore = new AccountsStore(
      new InMemoryStore(),
      new AsyncInMemoryStore()
    )

    const pullRequestStore = new PullRequestStore(
      new TestPullRequestDatabase(),
      repositoriesStore
    )

    const githubUserStore = new GitHubUserStore(db)
    const issuesStore = new IssuesStore(issuesDb)

    const repositoryStateManager = new RepositoryStateCache(repo =>
      githubUserStore!.getUsersForRepository(repo)
    )

    const apiRepositoriesStore = new ApiRepositoriesStore(accountsStore)

    const appStore = new AppStore(
      githubUserStore,
      new CloningRepositoriesStore(),
      issuesStore,
      statsStore,
      new SignInStore(),
      accountsStore,
      repositoriesStore,
      pullRequestStore,
      repositoryStateManager,
      apiRepositoriesStore
    )

    const dispatcher = new InMemoryDispatcher(
      appStore,
      repositoryStateManager,
      statsStore
    )

    repository = await setupEmptyRepository()

    defaultProps = {
      aheadBehind: null,
      remoteName: null,
      networkActionInProgress: false,
      progress: null,
      dispatcher,
      lastFetched: null,
      tipState: TipState.Valid,
      rebaseInProgress: false,
      branchWasRebased: false,
      repository,
    }
  })

  describe('render', () => {
    it(`when no remote found on repository`, () => {
      const { container } = render(<PushPullButton {...defaultProps} />)

      expect(container).toMatchSnapshot()
    })

    it(`when remote set but no tracking branch set`, () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when detached HEAD`, async () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        tipState: TipState.Detached,
      }
      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when detached HEAD and rebase in progress`, () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        tipState: TipState.Detached,
        rebaseInProgress: true,
      }
      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when unborn branch`, async () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        tipState: TipState.Unborn,
      }
      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when up to date with tracking branch`, () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 0, behind: 0 },
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when ahead of tracking branch`, () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 5, behind: 0 },
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when behind tracking branch`, () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 0, behind: 5 },
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when pull.rebase enabled`, () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 2, behind: 5 },
        pullWithRebase: true,
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when progress shown`, () => {
      const progress: IFetchProgress = {
        kind: 'fetch',
        remote: 'origin',
        value: 50,
        title: 'something',
      }

      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 0, behind: 5 },
        progress,
        networkActionInProgress: true,
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when progress shown but network action off`, () => {
      const progress: IFetchProgress = {
        kind: 'fetch',
        remote: 'origin',
        value: 50,
        title: 'something',
      }

      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 0, behind: 5 },
        progress,
        networkActionInProgress: false,
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it(`when last fetch known`, () => {
      const fiveHoursAgo = moment('2019-01-01')
        .subtract(5, 'hours')
        .toDate()

      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 0, behind: 0 },
        lastFetched: fiveHoursAgo,
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it('when branch is not published to GitHub remote', () => {
      const githubRepository = new Repository(
        repository.path,
        1,
        {
          dbID: 1,
          name: 'desktop',
          owner: {
            login: 'desktop',
            endpoint: 'https://api.github.com/',
            id: 1234,
            hash: 'some-other-hash',
          },
          private: false,
          htmlURL: 'https://github.com/desktop/desktop',
          cloneURL: 'https://github.com/desktop/desktop.git',
          defaultBranch: 'development',
          fullName: 'desktop/desktop',
          parent: null,
          endpoint: 'https://api.github.com/',
          fork: false,
          hash: 'some-value',
        },
        false
      )

      const props = {
        ...defaultProps,
        remoteName: 'origin',
        repository: githubRepository,
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })

    it('when branch can be force pushed', () => {
      const props = {
        ...defaultProps,
        remoteName: 'origin',
        aheadBehind: { ahead: 4, behind: 5 },
        branchWasRebased: true,
      }

      const { container } = render(<PushPullButton {...props} />)

      expect(container).toMatchSnapshot()
    })
  })
})
