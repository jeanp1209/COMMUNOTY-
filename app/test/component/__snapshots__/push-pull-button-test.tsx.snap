// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PushPullButton/> render when ahead of tracking branch 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 10 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M5 3L0 9h3v4h4V9h3L5 3z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Push origin
        </div>
        <div
          class="description"
        >
          Never fetched
        </div>
      </div>
      <div
        class="ahead-behind"
      >
        <span>
          5
          <svg
            aria-labelledby="octiconTitle octiconDescription"
            class="octicon"
            version="1.1"
            viewBox="0 0 6 16"
          >
            <title
              id="octiconTitle"
            />
            <desc
              id="octiconDescription"
            />
            <path
              d="M3 5L0 9h2v2h2V9h2L3 5z"
            />
          </svg>
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when behind tracking branch 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 10 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 7V3H3v4H0l5 6 5-6H7z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Pull origin
        </div>
        <div
          class="description"
        >
          Never fetched
        </div>
      </div>
      <div
        class="ahead-behind"
      >
        <span>
          5
          <svg
            aria-labelledby="octiconTitle octiconDescription"
            class="octicon"
            version="1.1"
            viewBox="0 0 6 16"
          >
            <title
              id="octiconTitle"
            />
            <desc
              id="octiconDescription"
            />
            <path
              d="M4 7V5H2v2H0l3 4 3-4H4z"
            />
          </svg>
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when branch can be force pushed 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 10 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M3 11H0l5-6 5 6H7v4H3v-4zM5 1l5 6H8.33L5 3 1.662 7H0l5-6z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Force push origin
        </div>
        <div
          class="description"
        >
          Never fetched
        </div>
      </div>
      <div
        class="ahead-behind"
      >
        <span>
          4
          <svg
            aria-labelledby="octiconTitle octiconDescription"
            class="octicon"
            version="1.1"
            viewBox="0 0 6 16"
          >
            <title
              id="octiconTitle"
            />
            <desc
              id="octiconDescription"
            />
            <path
              d="M3 5L0 9h2v2h2V9h2L3 5z"
            />
          </svg>
        </span>
        <span>
          5
          <svg
            aria-labelledby="octiconTitle octiconDescription"
            class="octicon"
            version="1.1"
            viewBox="0 0 6 16"
          >
            <title
              id="octiconTitle"
            />
            <desc
              id="octiconDescription"
            />
            <path
              d="M4 7V5H2v2H0l3 4 3-4H4z"
            />
          </svg>
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when branch is not published to GitHub remote 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 16 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 9H5l3-3 3 3H9v5H7V9zm5-4c0-.44-.91-3-4.5-3C5.08 2 3 3.92 3 6 1.02 6 0 7.52 0 9c0 1.53 1 3 3 3h3v-1.3H3c-1.62 0-1.7-1.42-1.7-1.7 0-.17.05-1.7 1.7-1.7h1.3V6c0-1.39 1.56-2.7 3.2-2.7 2.55 0 3.13 1.55 3.2 1.8v1.2H12c.81 0 2.7.22 2.7 2.2 0 2.09-2.25 2.2-2.7 2.2h-2V12h2c2.08 0 4-1.16 4-3.5C16 6.06 14.08 5 12 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Publish branch
        </div>
        <div
          class="description"
        >
          Publish this branch to GitHub
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when detached HEAD 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      disabled=""
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 16 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 9H5l3-3 3 3H9v5H7V9zm5-4c0-.44-.91-3-4.5-3C5.08 2 3 3.92 3 6 1.02 6 0 7.52 0 9c0 1.53 1 3 3 3h3v-1.3H3c-1.62 0-1.7-1.42-1.7-1.7 0-.17.05-1.7 1.7-1.7h1.3V6c0-1.39 1.56-2.7 3.2-2.7 2.55 0 3.13 1.55 3.2 1.8v1.2H12c.81 0 2.7.22 2.7 2.2 0 2.09-2.25 2.2-2.7 2.2h-2V12h2c2.08 0 4-1.16 4-3.5C16 6.06 14.08 5 12 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Publish branch
        </div>
        <div
          class="description"
        >
          Cannot publish detached HEAD
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when detached HEAD and rebase in progress 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      disabled=""
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 16 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 9H5l3-3 3 3H9v5H7V9zm5-4c0-.44-.91-3-4.5-3C5.08 2 3 3.92 3 6 1.02 6 0 7.52 0 9c0 1.53 1 3 3 3h3v-1.3H3c-1.62 0-1.7-1.42-1.7-1.7 0-.17.05-1.7 1.7-1.7h1.3V6c0-1.39 1.56-2.7 3.2-2.7 2.55 0 3.13 1.55 3.2 1.8v1.2H12c.81 0 2.7.22 2.7 2.2 0 2.09-2.25 2.2-2.7 2.2h-2V12h2c2.08 0 4-1.16 4-3.5C16 6.06 14.08 5 12 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Publish branch
        </div>
        <div
          class="description"
        >
          Rebase in progress
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when last fetch known 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 12 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M10.24 7.4a4.15 4.15 0 0 1-1.2 3.6 4.346 4.346 0 0 1-5.41.54L4.8 10.4.5 9.8l.6 4.2 1.31-1.26c2.36 1.74 5.7 1.57 7.84-.54a5.876 5.876 0 0 0 1.74-4.46l-1.75-.34zM2.96 5a4.346 4.346 0 0 1 5.41-.54L7.2 5.6l4.3.6-.6-4.2-1.31 1.26c-2.36-1.74-5.7-1.57-7.85.54C.5 5.03-.06 6.65.01 8.26l1.75.35A4.17 4.17 0 0 1 2.96 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Fetch origin
        </div>
        <div
          class="description"
        >
          <span>
            Last fetched 
            <span
              title="Monday, December 31, 2018 7:00 PM"
            >
              Dec 31, 2018
            </span>
          </span>
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when no remote found on repository 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 16 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 9H5l3-3 3 3H9v5H7V9zm5-4c0-.44-.91-3-4.5-3C5.08 2 3 3.92 3 6 1.02 6 0 7.52 0 9c0 1.53 1 3 3 3h3v-1.3H3c-1.62 0-1.7-1.42-1.7-1.7 0-.17.05-1.7 1.7-1.7h1.3V6c0-1.39 1.56-2.7 3.2-2.7 2.55 0 3.13 1.55 3.2 1.8v1.2H12c.81 0 2.7.22 2.7 2.2 0 2.09-2.25 2.2-2.7 2.2h-2V12h2c2.08 0 4-1.16 4-3.5C16 6.06 14.08 5 12 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Publish repository
        </div>
        <div
          class="description"
        >
          Publish this repository to GitHub
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when progress shown 1`] = `
<div>
  <div
    class="toolbar-button has-progress push-pull-button"
  >
    <button
      class="button-component"
      disabled=""
      type="button"
    >
      <div
        class="progress"
        style="transform: scaleX(1);"
      />
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon spin"
        version="1.1"
        viewBox="0 0 12 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M10.24 7.4a4.15 4.15 0 0 1-1.2 3.6 4.346 4.346 0 0 1-5.41.54L4.8 10.4.5 9.8l.6 4.2 1.31-1.26c2.36 1.74 5.7 1.57 7.84-.54a5.876 5.876 0 0 0 1.74-4.46l-1.75-.34zM2.96 5a4.346 4.346 0 0 1 5.41-.54L7.2 5.6l4.3.6-.6-4.2-1.31 1.26c-2.36-1.74-5.7-1.57-7.85.54C.5 5.03-.06 6.65.01 8.26l1.75.35A4.17 4.17 0 0 1 2.96 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          something
        </div>
        <div
          class="description"
        >
          Hang on…
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when progress shown but network action off 1`] = `
<div>
  <div
    class="toolbar-button has-progress push-pull-button"
  >
    <button
      class="button-component"
      disabled=""
      type="button"
    >
      <div
        class="progress"
        style="transform: scaleX(1);"
      />
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 12 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M10.24 7.4a4.15 4.15 0 0 1-1.2 3.6 4.346 4.346 0 0 1-5.41.54L4.8 10.4.5 9.8l.6 4.2 1.31-1.26c2.36 1.74 5.7 1.57 7.84-.54a5.876 5.876 0 0 0 1.74-4.46l-1.75-.34zM2.96 5a4.346 4.346 0 0 1 5.41-.54L7.2 5.6l4.3.6-.6-4.2-1.31 1.26c-2.36-1.74-5.7-1.57-7.85.54C.5 5.03-.06 6.65.01 8.26l1.75.35A4.17 4.17 0 0 1 2.96 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          something
        </div>
        <div
          class="description"
        >
          Hang on…
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when pull.rebase enabled 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 10 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 7V3H3v4H0l5 6 5-6H7z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Pull origin with rebase
        </div>
        <div
          class="description"
        >
          Never fetched
        </div>
      </div>
      <div
        class="ahead-behind"
      >
        <span>
          2
          <svg
            aria-labelledby="octiconTitle octiconDescription"
            class="octicon"
            version="1.1"
            viewBox="0 0 6 16"
          >
            <title
              id="octiconTitle"
            />
            <desc
              id="octiconDescription"
            />
            <path
              d="M3 5L0 9h2v2h2V9h2L3 5z"
            />
          </svg>
        </span>
        <span>
          5
          <svg
            aria-labelledby="octiconTitle octiconDescription"
            class="octicon"
            version="1.1"
            viewBox="0 0 6 16"
          >
            <title
              id="octiconTitle"
            />
            <desc
              id="octiconDescription"
            />
            <path
              d="M4 7V5H2v2H0l3 4 3-4H4z"
            />
          </svg>
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when remote set but no tracking branch set 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 16 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 9H5l3-3 3 3H9v5H7V9zm5-4c0-.44-.91-3-4.5-3C5.08 2 3 3.92 3 6 1.02 6 0 7.52 0 9c0 1.53 1 3 3 3h3v-1.3H3c-1.62 0-1.7-1.42-1.7-1.7 0-.17.05-1.7 1.7-1.7h1.3V6c0-1.39 1.56-2.7 3.2-2.7 2.55 0 3.13 1.55 3.2 1.8v1.2H12c.81 0 2.7.22 2.7 2.2 0 2.09-2.25 2.2-2.7 2.2h-2V12h2c2.08 0 4-1.16 4-3.5C16 6.06 14.08 5 12 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Publish branch
        </div>
        <div
          class="description"
        >
          Publish this branch to the remote
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when unborn branch 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      disabled=""
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 16 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M7 9H5l3-3 3 3H9v5H7V9zm5-4c0-.44-.91-3-4.5-3C5.08 2 3 3.92 3 6 1.02 6 0 7.52 0 9c0 1.53 1 3 3 3h3v-1.3H3c-1.62 0-1.7-1.42-1.7-1.7 0-.17.05-1.7 1.7-1.7h1.3V6c0-1.39 1.56-2.7 3.2-2.7 2.55 0 3.13 1.55 3.2 1.8v1.2H12c.81 0 2.7.22 2.7 2.2 0 2.09-2.25 2.2-2.7 2.2h-2V12h2c2.08 0 4-1.16 4-3.5C16 6.06 14.08 5 12 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Publish branch
        </div>
        <div
          class="description"
        >
          Cannot publish unborn HEAD
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`<PushPullButton/> render when up to date with tracking branch 1`] = `
<div>
  <div
    class="toolbar-button push-pull-button"
  >
    <button
      class="button-component"
      type="button"
    >
      <svg
        aria-labelledby="octiconTitle octiconDescription"
        class="octicon icon"
        version="1.1"
        viewBox="0 0 12 16"
      >
        <title
          id="octiconTitle"
        />
        <desc
          id="octiconDescription"
        />
        <path
          d="M10.24 7.4a4.15 4.15 0 0 1-1.2 3.6 4.346 4.346 0 0 1-5.41.54L4.8 10.4.5 9.8l.6 4.2 1.31-1.26c2.36 1.74 5.7 1.57 7.84-.54a5.876 5.876 0 0 0 1.74-4.46l-1.75-.34zM2.96 5a4.346 4.346 0 0 1 5.41-.54L7.2 5.6l4.3.6-.6-4.2-1.31 1.26c-2.36-1.74-5.7-1.57-7.85.54C.5 5.03-.06 6.65.01 8.26l1.75.35A4.17 4.17 0 0 1 2.96 5z"
        />
      </svg>
      <div
        class="text"
      >
        <div
          class="title"
        >
          Fetch origin
        </div>
        <div
          class="description"
        >
          Never fetched
        </div>
      </div>
    </button>
  </div>
</div>
`;
